### PR TITLE
feat/gate io connector update

### PR DESCRIPTION
**Pull request by origliante**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Gate io exchange connector refactoring to recent team standards;
Proposal for exchange_base_v2 to raise code reuse across connector exchanges;
Refactoring of kucoin and binance connectors to inherit from exchange_base_v2;
Proposal of `integration_test_tool.py` in `test/connector` to aid manual testing with a live exchange;

**Tests performed by the developer**:
Unit tests for binance and kucoin;
Partial unit tests for gate io (WIP unit tests for gate_io_exchange module);
Live interactive testing for gate io;
Pure mm for gate io with spreads of 0.5% and 10s refresh time;

**Tips for QA testing**:
Please start testing from binance and kucoin;
And if possible give a try to `integration_test_tool.py`: the idea is to develop the tool in collaboration with QA so that we can automate common use cases and live test them on the real exchange to reduce manual work and reduce risk of human error even further. Can be run like: 
```
conda activate hummingbot
PYTHONPATH=. python test/connector/integration_test_tool.py
```
(press enter to see the prompt)

### CODE REVIEW

**Comment by RC-13** [comment]
@origliante I started testing the PR and currently getting the following issues:
1. When running `integration_test_tool.py`, the following error comes up:
```python
(hummingbot) ubuntu@ip-172-31-29-21:~/gate-update$ PYTHONPATH=. python test/connector/integration_test_tool.py
Traceback (most recent call last):
  File "test/connector/integration_test_tool.py", line 18, in <module>
    from certs import creds
ModuleNotFoundError: No module named 'certs'
```
2. When starting a strategy using the Binance connector, the strategy fails to start with the following error in the logs:
```python
2022-04-25 08:16:57,199 - 6630 - hummingbot.connector.exchange_base_v2 - NETWORK - Unexpected error while fetching account updates.
Traceback (most recent call last):
  File "/Users/clintg/testing/gate/hummingbot/connector/exchange_base_v2.py", line 711, in _status_polling_loop
    await self._status_polling_loop_fetch_updates()
AttributeError: 'BinanceExchange' object has no attribute '_status_polling_loop_fetch_updates'
```
3. And if we check the status, it doesn't display the correct base asset balance
image
image
4. Gate_io currently accepts invalid API key
image
image
```python
2022-04-25 02:09:42,807 - 24347 - hummingbot.connector.exchange_base_v2 - WARNING - Could not fetch balance update from Gate_io
2022-04-25 02:10:24,792 - 24347 - hummingbot.connector.exchange_base_v2 - NETWORK - Unexpected error while fetching balance update - Error executing request GET spot/accounts. HTTP status is 401. Error: {"label":"INVALID_KEY","message":"Invalid key provided"}
Traceback (most recent call last):
  File "/home/ubuntu/gate-update/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 182, in _update_balances
    account_info = await self._api_get(
  File "/home/ubuntu/gate-update/hummingbot/connector/exchange_base_v2.py", line 766, in _api_get
    return await self._api_request(*args, **kwargs)
  File "/home/ubuntu/gate-update/hummingbot/connector/exchange_base_v2.py", line 788, in _api_request
    return await self.web_utils.api_request(
  File "/home/ubuntu/gate-update/hummingbot/connector/exchange/gate_io/gate_io_web_utils.py", line 145, in api_request
    raise IOError(
OSError: Error executing request GET spot/accounts. HTTP status is 401. Error: {"label":"INVALID_KEY","message":"Invalid key provided"}
```
[binance-pmm.log](https://github.com/CoinAlpha/hummingbot/files/8551205/binance-pmm.log)

**Comment by Deepali6997** [comment]
**gate.io**

1. When pure market making strategy is created and started, it does not starts and says **"Market connectors are not ready yet."** Tried deploying strategy in master branch with same configs and it creates orders successfully without error. 

> PR198 branch


Screenshot 2022-04-25 at 12 01 23 PM

logs file : [logs_gateio_pmm_PR198.log](https://github.com/CoinAlpha/hummingbot/files/8552132/logs_gateio_pmm_PR198.log)

> MASTER branch


Screenshot 2022-04-25 at 12 05 44 PM

logs file : [logs_gateio_pmm_master.log](https://github.com/CoinAlpha/hummingbot/files/8552161/logs_gateio_pmm_master.log)

csv file : [trades_gateio_pmm_master.csv](https://github.com/CoinAlpha/hummingbot/files/8552158/trades_gateio_pmm_master.csv)



**KUCOIN**

1. When we run config rate_oracle_source_kucoin , the balance command then starts showing zero value for assets and gross balance as zero for KuCoin exchange. In master branch too it is showing same behaviour

>  PR 198 branch


Before running config rate_oracle_source_kucoin 

Screenshot 2022-04-25 at 12 12 22 PM

After running config rate_oracle_source_kucoin ; it starts showing zero allocated and total %

Screenshot 2022-04-25 at 12 14 06 PM

> MASTER branch

Screenshot 2022-04-25 at 12 18 23 PM

File: hummingbot/connector/exchange/gate_io/gate_io_utils.py [hunk 1]
```diff
@@ -1,176 +1,3 @@
...
-                  is_secure=True,
-                  is_connect_key=True),
-    "gate_io_secret_key":
-        ConfigVar(key="gate_io_secret_key",
-                  prompt=f"Enter your {CONSTANTS.EXCHANGE_NAME} secret key >>> ",
-                  required_if=using_exchange("gate_io"),
-                  is_secure=True,
-                  is_connect_key=True),
-}
+from .gate_io_constants import *  # noqa: F403 F401
```
Comment (aarmoa) [inline hummingbot/connector/exchange/gate_io/gate_io_utils.py:3]: This implementation is not clear and can confuse the community. We should keep the utils module as it was (no need to unify it with the constants)

**Review by aarmoa — CHANGES_REQUESTED** [review]

File: hummingbot/connector/exchange_base_v2.py [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+        the time of the client to be the same as the time in the exchange.
+        Executes when the _poll_notifier event is enabled by the `tick` function.
+        """
+        while True:
+            try:
+                await self._poll_notifier.wait()
+                await self._update_time_synchronizer()
+
+                # the following method is implementation-specific
+                await self._status_polling_loop_fetch_updates()
```
Comment (aarmoa) [inline hummingbot/connector/exchange_base_v2.py:711]: `_status_polling_loop_fetch_updates` should be implemented in this class as an abstract method

File: hummingbot/connector/exchange/binance/binance_exchange.py [hunk 3]
```diff
@@ -1,47 +1,47 @@
...
-from hummingbot.core.web_assistant.rest_assistant import RESTAssistant
-from hummingbot.logger import HummingbotLogger
+from hummingbot.core.utils.async_utils import safe_gather
+
 
-s_logger = None
-s_decimal_0 = Decimal(0)
-s_decimal_NaN = Decimal("nan")
+class BinanceExchange(ExchangeBaseV2):
+    DEFAULT_DOMAIN = CONSTANTS.DEFAULT_DOMAIN
```
[the remark this answers is not in the document]
Comment (origliante) [reply hummingbot/connector/exchange/binance/binance_exchange.py:26]: If I redefine both _logger and def logger() in the implementation class,
by logging from supported_order_types() - which is in the Base V2, I still get:

INFO	[exchange_base_v2.py]	[supported_order_types] |	test

But all logging from methods that are redefined in the implementation class, log correctly even if I do not redefine anything:

DEBUG	[gate_io_exchange.py]	[_status_polling_loop_fetch_updates] |	Running _status_polling_loop_fetch_updates() at 1650973044.478204

**Comment by petioptrv** [comment]
[ch22462]

**Comment by origliante** [comment]
@CoinAlpha/qa-team hey guys pls give it another round, the reported issues should be resolved

**Comment by Deepali6997** [comment]
Test Update

- Can confirm that reported issues are all resolved
- There was one issue that was spotted and i.e. **when we're creating a PMM strategy, token trading pairs are not getting auto-populated when typing the trading pair or hitting the tab**

Screenshot 2022-04-28 at 11 29 12 AM

**Comment by RC-13** [comment]
Confirm that the issue with the trading pair not getting displayed is fixed in the latest commit. Although another error comes up whenever a strategy is started  (same error is captured when running the integration_test_tool.py):
image
```python
2022-05-03 05:22:02,763 - 18443 - hummingbot.connector.exchange.gate_io.gate_io_api_order_book_data_source - ERROR - Unexpected error while parsing ws trades message {'time': 1651555322, 'channel': 'spot.trades', 'event': 'subscribe', 'result': {'status': 'success'}}.
Traceback (most recent call last):
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_api_order_book_data_source.py", line 287, in listen_for_trades
    pair: str = await self.trading_pair_associated_to_exchange_symbol(trade_data.get("currency_pair", None))
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_api_order_book_data_source.py", line 172, in trading_pair_associated_to_exchange_symbol
    return symbol_map[symbol]
  File "/home/ubuntu/miniconda3/envs/gate/lib/python3.8/site-packages/bidict/_base.py", line 530, in __getitem__
    return self._fwdm[key]
KeyError: None
```
```python
2022-05-03 05:22:02,764 - 18443 - hummingbot.connector.exchange.gate_io.gate_io_api_order_book_data_source - ERROR - Unexpected error while parsing ws order book message {'time': 1651555322, 'channel': 'spot.order_book_update', 'event': 'subscribe', 'result': {'status': 'success'}}.
Traceback (most recent call last):
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_api_order_book_data_source.py", line 327, in listen_for_order_book_diffs
    timestamp: float = (order_book_data["t"]) * 1e-3
KeyError: 't'
```
```python
2022-05-03 05:22:03,054 - 18443 - hummingbot.connector.exchange_py_base - ERROR - Unexpected error in user stream listener loop.
Traceback (most recent call last):
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 300, in _user_stream_event_listener
    self._process_order_message(order_msg)
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 352, in _process_order_message
    client_order_id = str(order_msg.get("text", ""))
AttributeError: 'str' object has no attribute 'get'
```
```python
2022-05-03 05:22:08,056 - 18443 - hummingbot.connector.exchange_py_base - ERROR - Unexpected error in user stream listener loop.
Traceback (most recent call last):
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 297, in _user_stream_event_listener
    self._process_trade_message(trade_msg)
  File "/home/ubuntu/gate/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 377, in _process_trade_message
    client_order_id = client_order_id or str(trade["text"])
TypeError: string indices must be integers
```
Despite getting the error, the strategy continue to work as expected.
[logs_conf_pure_mm_1.log](https://github.com/CoinAlpha/hummingbot/files/8608379/logs_conf_pure_mm_1.log)

**Comment by gerald-hb** [comment]
Latest commit looks good to me. I can confirm that the previous error no longer occurs.

Test summary.
1. clone and install the featured branch :white_check_mark:
2. Add API keys and check out balance command :white_check_mark:
3. Create pure market making strategy, check status/--live :white_check_mark:
4. Buy and sell orders were created and canceled accordingly :white_check_mark:
5. Cross check trading fees in history, orderbook and .csv file (data folder).  :white_check_mark:

**Comment by aarmoa** [comment]
Hello @CoinAlpha/qa-team. All refactorings have been applied to the PR. Could you perform a quick test again for the connectors (Binance, Kucoin, Gateio)?

**Comment by Deepali6997** [comment]
test update for Kucoin connector

1. Cloning and installation of feature branch and checking behaviour for API incorrect and correct credentials
2. Creation of strategy and checking status --live, ticker --live, orderbook --live, history and history --verbose commands
3. Checking the strategy for fully filled, partially filled orders and fee calculation
4. Checking filled orders in datadog under graph (hummingbot.filled_usdt_volume)
5. Checking slow and fast refresh rate 
6. Checking for multiple bots and checking the hangingorder behavior

Can confirm the above mentioned points are working fine.

will update soon for other two connectors Binance and gate.io

**Comment by gerald-hb** [comment]
Run the remaining test for gate and Binance. Changes looks good to me. 

Test summary

1.clone and install the featured branch ✅
2.Add API keys and check out balance command ✅
3.Create pure market making strategy, check status/--live, ticker --live, orderbook --live, history and history --verbose commands ✅
4.Buy and sell orders were created and canceled accordingly ✅
5.Checking slow and fast refresh rate✅
5.Cross check trading fees in history, orderbook and .csv file (data folder). ✅

**Review by RC-13 — APPROVED** [review]
Tested and confirmed working, LGTM!

Test summary:
- Install and start the feature branch without any issue
- Connect Binance, Gate_io, and KuCoin API key. Gate_io did have an issue with accepting invalid API keys which had been fixed within the commits.
- Checked the balance and successfully returns and displays accurate balance in each exchange.
- Created different pure market-making strategies using Binance, Gate_io, and KuCoin as exchanges. A minor issue was seen during the strategy creation specific to Gate_io which has been sorted within the commits.
- Started the strategy, initially, errors occurred when starting any strategy which was confirmed fixed in the latest commit.
- Confirmed that the order creation and cancellation follow the specified parameter set in the strategy.
- Stopped and restarted the strategy multiple times to verify that orders are properly getting canceled.
- Tested on different refresh rates to confirm that there are no orders left out in the exchange.
- Filled orders are confirmed and recorded correctly and fees match the fee record in the exchange.
- Checked the orders and prices to make sure that they are synchronized with the exchange.
- Run multiple bots using the same exchange and market to confirm that there are no orderID conflicts.
- Add the API key by creating a python file and testing the integration_test_tool.py. Confirms that the tool is running and capturing issues if there are any.

**Review by petioptrv — CHANGES_REQUESTED** [review]
I like the changes. The new code is very clean, and it simplifies new implementations quite a bit. I've left some comments in the code.

One general comment that I have is that I don't see the need to make the interface abstract through private versions of the interface functions. I think the interface function itself should be marked as abstract. That way, it is immediately apparent to the implementer that this is something they should define, rather than having to follow the function calls to the private function that is finally abstract. Also, very very minor, but this removes the overhead of yet another function call.

Also, there are many methods that raise `NotImplementedError`, but are not marked as abstract. To be "pythonic", we should mark those with `@abstractmethod` decorator, and have them `pass` instead. Like:
```Python
class MyABCClass(ABC):
    @abstractmethod
    def my_abc_method(self):
        pass
```
Only if we inherit from `ABC` and decorate with`abstractmethod` will Python raise when the abstract methods are not implemented on class non-abstract class definition. Also, if both of those conditions are met, then the body of the function has no effect.

File: hummingbot/client/command/start_command.py [hunk 4]
```diff
@@ -3,8 +3,7 @@
 import threading
 import time
 from os.path import dirname, exists, join
-from typing import Any, Callable, Dict, List, Optional
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/client/command/start_command.py:6]: I thought that isort was executed automatically now before any commit

File: hummingbot/connector/exchange_py_base.py [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+from hummingbot.core.data_type.user_stream_tracker import UserStreamTracker
+from hummingbot.core.data_type.user_stream_tracker_data_source import UserStreamTrackerDataSource
+from hummingbot.core.network_iterator import NetworkStatus
+from hummingbot.core.utils.async_utils import safe_ensure_future, safe_gather
+from hummingbot.core.web_assistant.connections.data_types import RESTMethod
+from hummingbot.core.web_assistant.web_assistants_factory import WebAssistantsFactory
+from hummingbot.logger import HummingbotLogger
+
+
+class ExchangePyBase(ExchangeBase):
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/connector/exchange_py_base.py:33]: I tried this but is not possible. Everything breaks if you try to use multiple inheritance in a class that inherits from a Cython class.

File: hummingbot/client/command/start_command.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (petioptrv) [reply hummingbot/client/command/start_command.py:6]: It does, and it currently fails on the pipeline. My guess is that those changes were pushed before isort was merged into dev. Or maybe pre-commit is not installed locally

File: hummingbot/connector/exchange_py_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (petioptrv) [reply hummingbot/connector/exchange_py_base.py:33]: I may be missing something, but this minimal example works on my end:
```Python
    def test(self):
        pass


class TestImpl(Test):
    def test(self):
        print("test")


if __name__ == '__main__':
    try:
        t = Test()
    except TypeError as e:
        print(e)

    t = TestImpl()
    t.test()
```

File: hummingbot/client/command/start_command.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/client/command/start_command.py:6]: Applying the change in the next commit

File: test/connector/integration_test_tool.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        print("ORDERS: ", [o.to_json() for o in self.exchange.in_flight_orders.values()])
+        await self.exchange._update_order_status()
+
+    async def tests(self):
+        global async_input
+        prompt = Prompt()
+        async_input = functools.partial(prompt, end='', flush=True)
+
+        global ENABLE_CONNECTION_TRACING
+        # ENABLE_CONNECTION_TRACING = True
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:238]: Removing the comment in the next commit

File: test/connector/integration_test_tool.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            await asyncio.sleep(5)
+
+            r = await self.exchange.check_network()
+            print(f'\n{r}')
+            if r == NetworkStatus.CONNECTED:
+                print("Exchange status: ", self.exchange.status_dict)
+                # print("Trading pair symbol map from OB DS: ", self.exchange._orderbook_ds._trading_pair_symbol_map)
+                # api_keys = { "gate_io_api_key": creds.k, "gate_io_secret_key": creds.s, }
+                # Can be used to test for invalid keys
+                # await UserBalances.instance().add_exchange(self.exchange.name, **api_keys)
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:251]: Removing the commented print in the next commit

File: test/connector/integration_test_tool.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                min_base_amount_increment=Decimal(str(0.000001)),
+                max_order_size=Decimal(str(0.001)),
+            )
+        }
+        print("Trading rules: ", self.exchange._trading_rules)
+
+    async def start(self):
+        self.set_trading_pair("ETH", "USDT")
+
+        # self.exchange = get_binance(self)
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:213]: Removing the comment in the next commit

File: test/connector/integration_test_tool.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+def get_gate_io(ec):
+    exchange = GateIoExchange(
+        gate_io_api_key=creds.k,
+        gate_io_secret_key=creds.s,
+        trading_pairs=[ec.pair],
+    )
+    # exchange.ORDERBOOK_DS_CLASS._trading_pair_symbol_map = {
+    #    exchange.DEFAULT_DOMAIN: bidict({f"{ec.base}{ec.quote}": ec.pair})
+    # }
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:140]: Removing the comment in the next commit

File: test/connector/integration_test_tool.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+
+def get_binance(ec):
+    exchange = BinanceExchange(
+        binance_api_key=creds.k,
+        binance_api_secret=creds.s,
+        trading_pairs=[ec.pair],
+    )
+    # exchange.ORDERBOOK_DS_CLASS._trading_pair_symbol_map = {
+    #    "com": bidict({f"{ec.base}{ec.quote}": ec.pair})
+    # }
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:128]: Removing the comment in the next commit

File: test/connector/integration_test_tool.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+import time
+import logging
+import asyncio
+from distutils.util import strtobool
+import functools
+
+from decimal import Decimal
+import aiohttp
+
+# from hummingbot.user.user_balances import UserBalances
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/connector/integration_test_tool.py:11]: Removing the comment in the next commit

File: hummingbot/connector/exchange/kucoin/kucoin_exchange.py [hunk 12]
```diff
@@ -1,57 +1,30 @@
...
-    UPDATE_ORDERS_INTERVAL = 10.0
-    SHORT_POLL_INTERVAL = 5.0
-    LONG_POLL_INTERVAL = 120.0
 
-    @classmethod
-    def logger(cls) -> HummingbotLogger:
-        if cls._logger is None:
-            cls._logger = logging.getLogger(__name__)
-        return cls._logger
+    web_utils = web_utils
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/connector/exchange/kucoin/kucoin_exchange.py:27]: I am planning on changing the web_utils reference in the OKX PR, when both changes are merged together. I don't think it is a good idea to work on reifying the utilities concept as part of this PR.

File: hummingbot/connector/exchange/gate_io/gate_io_web_utils.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        self.http_status = error_payload.get('status')
+        if isinstance(error_payload, dict):
+            self.error_message = error_payload.get('error', error_payload).get('message', error_payload)
+            self.error_label = error_payload.get('error', error_payload).get('label', error_payload)
+        else:
+            self.error_message = error_payload
+            self.error_label = error_payload
+
+
+class GateIoWebsocket:
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply hummingbot/connector/exchange/gate_io/gate_io_web_utils.py:179]: Removing the class in the next commit

**Review by petioptrv — APPROVED** [review]
The changes LGTM!

**Comment by aarmoa** [comment]
Closing this PR after creating its equal in the foundation repository (https://github.com/hummingbot/hummingbot/pull/5313)

**Pull request closed without merging**